### ISSUE TITLE
countUsed() functionality broken

### ISSUE THREAD

**Issue by johnda102** [UCATLAS/xAODAnaHelpers#366]
The base class has a registerClass function that is meant to keep track of how many instances of a class have been created.  

It seems this is not active.  Im creating multiple muonSelectors and I see this->countUsed() is always 0. 

Searching the c++ I never see AH::Algorithm::registerClass invoked (for any class). 

Has anyone used this functionality before ?  

It looks like 
ElectronSelector.cxx
MuonSelector.cxx
and 
TauSelector.cxx 

are accessing countUsed() function.

**Comment by kratsg** [comment]
Does @mmilesi want to take this over? Otherwise, I know how to generally implement the auto-registration. We need to make sure xAH_run.py is general to use with other people's scripts (standard EL algos) without relying on having to register manually.

**Comment by johnda102** [comment]
I would suggest we change things to give the selectors responsibility for making the cutflow histograms and adding them to the output file,  Instead of retrieving the existing histograms from BasicEventSelection. 

The selector can then name the cutflow histograms based on their given m_name.

This would remove the need to keep track of how make instances have been created and would allow support of more then 2 instances.

**Comment by kratsg** [comment]
@johnda102 - so you would support having something like:

- `m_doCutflow` as a boolean
- `m_name` being set differently for each algorithm in the case that a cutflow needs to be made?

**Comment by johnda102** [comment]
Here im talking about the object level cut flows.

So Basically remove the lines 
```c++
  m_mu_cutflowHist_1     = new TH1D("cutflow_muons_1", "cutflow_muons_1", 1, 1, 2);
  m_mu_cutflowHist_1->SetBit(TH1::kCanRebin);
  m_mu_cutflowHist_2     = new TH1D("cutflow_muons_2", "cutflow_muons_2", 1, 1, 2);
  m_mu_cutflowHist_2->SetBit(TH1::kCanRebin);
```
in BasicEventSelection

and have in the MuonSelector something like:

``` (fake c++)
  m_mu_cutflowHist     = new TH1D("cutflow_muons_"+m_name, "cutflow_muons_"+m_name, 1, 1, 2);
  m_mu_cutflowHist->SetBit(TH1::kCanRebin);
  // get the cutflow directory from BasicEventSelctor and set m_mu_cutFlowHist to it 
```

This way we can configure and arbitrary number of MuonSelectors and they will all have thier own muon level cutflow with an assocated name (isnstead of just "_1" and "_2")

**Comment by kratsg** [comment]
Ok, I'll take care of it this weekend.

**Comment by mmilesi** [comment]
Hi @johnda102 ,

yes, that implementation I made was kind of a draft, and the way it worked for me was to manually register the algorithms configured in the steering macro. See this for example:

https://github.com/mmilesi/HTopMultilepAnalysis/blob/master/util/test-HTopMultilep.cxx#L209

However, I just moved to the python configuration scheme you also use, and already had a brief chat with @kratsg on how to make this "algorithm bookkeeping" more general and reliable.

Thanks Giordon for taking care of this!

Marco
